Updates to AST generation

x-tyger opened the pull request:
Some changes to encode some additional information (especially around types) in the AST.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/elliotchance/c2go/161)
<!-- Reviewable:end -->

elliotchance commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[util/goast.go, line 41 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-KlpxvUubJM5wnlyutMU:-KlpxvUubJM5wnlyutMV:bgctu67) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L41)):*
> ```Go
> 	// The identifier may start with zero or more "[]" since types are used as
> 	// function names.
> 	return regexp.MustCompile(`^(\[\d*\])*[a-zA-Z_][a-zA-Z0-9_.]*$`).
> ```

Thanks to your changes you should be able to make this a more strict (sane) check for just alphanumeric chars.

---

*[util/goast.go, line 52 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-Klpy01W7JjG-IYts04F:-Klpy01W7JjG-IYts04G:blva7ev) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L52)):*
> ```Go
> 	}
> 
> 	return regexp.MustCompile(`^\**(\[\])*[a-zA-Z_][a-zA-Z0-9_.]*$`).
> ```

Can this also be fixed with your change? Put all types through `typeToExpr` so that we get proper AST instead of the dodgy way I've been doing it up until now?

---

*[util/goast.go, line 83 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-Klpx-Gfvdv6SSqGGGfH:-Klpx-Gfvdv6SSqGGGfI:bqhtarx) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L83)):*
> ```Go
> 		return &goast.ArrayType{
> 			Elt: typeToExpr(match[2]),
> 			// TODO: Should use NewIntLit, but it doesn't seem right to
> ```

Remove the "TODO" keyword since this code does not need to be fixed up - even if seems odd as you say.

---

*[util/goast.go, line 97 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-KlpxaY414ss-vgNh3E8:-KlpxaY414ss-vgNh3E9:bwhnv3m) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L97)):*
> ```Go
> 		i := strings.IndexByte(t, '.')
> 		return &goast.SelectorExpr{
> 			X:   goast.NewIdent(t[0:i]),
> ```

`X` could be another expression like: `foo().bar`, maybe best to do?

```go
X:   typeToExpr(t[0:i]),
```

---

*[util/goast.go, line 211 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-KlpvwlPC7PalTC5j_Qk:-KlpvwlPC7PalTC5j_Ql:b-sdh3pg) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L211)):*
> ```Go
> 	return &goast.BasicLit{
> 		Kind:  token.FLOAT,
> 		Value: strconv.FormatFloat(value, 'E', -1, 64),
> ```

I've run into an issue where very large/tiny floats can be formatted incorrectly, see https://github.com/elliotchance/c2go/blob/master/transpiler/literals.go#L20-L31

Can you please add a few unit tests to make sure these values are correct: `1.23e300`, `1.23e-300` and `1.23`. Or better, fix both of these cases by creating a function that formats floats correctly based on their magnitude.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/161)*
<!-- Sent from Reviewable.io -->

x-tyger commented:
Review status: 7 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[util/goast.go, line 41 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-KlpxvUubJM5wnlyutMU:-KlqZKseDTzuDU0exrnC:b-wvltx3) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L41)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Thanks to your changes you should be able to make this a more strict (sane) check for just alphanumeric chars.
</blockquote></details>

Agreed. Changed to just allow valid identifiers.

---

*[util/goast.go, line 52 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-Klpy01W7JjG-IYts04F:-KlqZaBqHO4MbEylYwur:bq9cej5) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L52)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Can this also be fixed with your change? Put all types through `typeToExpr` so that we get proper AST instead of the dodgy way I've been doing it up until now?
</blockquote></details>

I believe you are correct. I've removed this validation. If anything is missed it should get caught when the recursion of `typeToExpr` boils down to `NewIdent`

---

*[util/goast.go, line 83 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-Klpx-Gfvdv6SSqGGGfH:-KlqZqtrQ9bXUqY3Wh0x:b-896fix) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L83)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

Remove the "TODO" keyword since this code does not need to be fixed up - even if seems odd as you say.
</blockquote></details>

Done.

---

*[util/goast.go, line 97 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-KlpxaY414ss-vgNh3E8:-KlqZwSto-6C7cQqzEF7:b-pbdqz3) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L97)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

`X` could be another expression like: `foo().bar`, maybe best to do?

```go
X:   typeToExpr(t[0:i]),
```
</blockquote></details>

Agreed.

---

*[util/goast.go, line 211 at r1](https://reviewable.io:443/reviews/elliotchance/c2go/161#-KlpvwlPC7PalTC5j_Qk:-Klq_536KIHJ3uZ-PlVn:b-guabdi) ([raw file](https://github.com/elliotchance/c2go/blob/3c86eed2eacf58eb61a4ed58996b2289fa99eb9f/util/goast.go#L211)):*
<details><summary><i>Previously, elliotchance (Elliot Chance) wrote…</i></summary><blockquote>

I've run into an issue where very large/tiny floats can be formatted incorrectly, see https://github.com/elliotchance/c2go/blob/master/transpiler/literals.go#L20-L31

Can you please add a few unit tests to make sure these values are correct: `1.23e300`, `1.23e-300` and `1.23`. Or better, fix both of these cases by creating a function that formats floats correctly based on their magnitude.
</blockquote></details>

Done. Let me know if you think using 'g' in FormatFloat looks better to you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/161)*
<!-- Sent from Reviewable.io -->

elliotchance commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/elliotchance/c2go/161)*
<!-- Sent from Reviewable.io -->
